File: dbxml/examples/java/misc/Compression.java
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle. All rights reserved.
 *
 *
 *******
 *
 * Compression API
 *
 * A very simple Berkeley DB XML program that demonstrates
 * compression API usage.
 *
 * This program demonstrates:
 *  Registration of a custom compression class in XmlManager
 *  Compression configuration for XmlContainer
 *  Turn off default compression
 *  Custom compression class implementation
 *
 *  MyCompression is an subclass of XmlCompression, it demonstrates
 *  compression using JDK's ZLIB compression support. It is a very 
 *  simple implementation and not intended to be particularly efficient.
 *  
 *  Compression only works with whole document containers.
 *  User compression must always be registered with the manager in
 *  order to open/create the container.
 *  
 */

package misc;

import com.sleepycat.dbxml.*;

/* 
 * MyCompression implements the XmlCompression interface: required
 * methods are compress() and decompress(). The two methods are required 
 * to return true for success and false for error. This class 
 * must be thread-safe in multi-thread environment. 
 * 
 */
class MyCompression extends XmlCompression {

    public boolean compress(XmlTransaction txn, XmlData source, XmlData dest) {

	try {
	    // Get the data to compress
	    byte[] src = source.get_data();

	    // Use JDK's ZLIB compress 
	    java.util.zip.Deflater compressor = 
		new java.util.zip.Deflater();
	    compressor.setInput(src);
	    compressor.finish();

	    java.io.ByteArrayOutputStream bos = 
		new java.io.ByteArrayOutputStream(src.length);

	    // Compress the data
	    byte[] buf = new byte[1024];
	    while (!compressor.finished()) {
		int count = compressor.deflate(buf);
		bos.write(buf, 0, count);
	    }
	    bos.close();

	    byte[] data = bos.toByteArray();

	    // Set the compressed data
	    dest.set(data);

	} catch (Exception e) {
	    // If any exception, return false
	    return false;
	}

	// Successful return true
	return true;
    }

    public boolean decompress(XmlTransaction txn, XmlData source, XmlData dest) {
		
        try {
	    // Get the data to decompress
	    byte[] src = source.get_data();

	    // Use JDK's ZLIB decompress
	    java.util.zip.Inflater decompressor = 
		new java.util.zip.Inflater();
	    decompressor.setInput(src);

	    java.io.ByteArrayOutputStream bos = 
		new java.io.ByteArrayOutputStream(src.length);

	    // Decompress the data
	    byte[] buf = new byte[1024];
	    while (!decompressor.finished()) {
		int count = decompressor.inflate(buf);
		bos.write(buf, 0, count);
	    }
	    bos.close();

	    byte[] data = bos.toByteArray();

	    // Set the decompressed data
	    dest.set(data);

	} catch (Exception e) {
	    // If any exception, return false
	    return false;
	}

	// Successful return true
	return true;
    }
}

public class Compression {

    public static void main(String[] args) throws Throwable {

	String containerName1 = "compressionContainer1.dbxml";
	String containerName2 = "compressionContainer2.dbxml";
	String docName = "doc1.xml";
	String content = "<root><a></a></root>";
	String compressionName = "myCompression";
		
	try {

	    XmlManager mgr = new XmlManager();
			
	    // Start with fresh containers
	    if (mgr.existsContainer(containerName1) != 0)
		mgr.removeContainer(containerName1);

	    if (mgr.existsContainer(containerName2) != 0)
		mgr.removeContainer(containerName2);
			
	    // Turn off the default compression			 
            XmlContainerConfig containerConfig = new XmlContainerConfig();
	    // Set the container type WholedocContainer 
	    // and turn off default compression
	    containerConfig.setContainerType(XmlContainer.WholedocContainer);
	    containerConfig.setCompression(XmlContainerConfig.NO_COMPRESSSION);
			
	    // Create container
	    XmlContainer cont = mgr.createContainer(containerName1, 
						    containerConfig);
			
	    XmlDocument doc = mgr.createDocument();
	    doc.setName(docName);
	    doc.setContent(content);
	    cont.putDocument(doc);
			
	    cont.delete();
			
	    // Open the container 
	    cont = mgr.openContainer(containerName1);
			
	    // Verify the content
	    System.out.println("Content of the document: " 
			       + cont.getDocument(docName).getContentAsString());
			
	    cont.delete();
			
			
	    // Following code demonstrates: 
	    //   compression with user's compression class
			 			
	    // Register user's compression object into XmlManager
	    MyCompression compressor = new MyCompression();
	    mgr.registerCompression(compressionName, compressor);

	    containerConfig = new XmlContainerConfig();
	    containerConfig.setContainerType(XmlContainer.WholedocContainer);
			
	    // Set XmlContainerConfig custom compression
	    containerConfig.setCompression(compressionName);
			
	    cont = mgr.createContainer(containerName2, containerConfig);
			
	    // Put a document in
	    doc = mgr.createDocument();
	    doc.setName(docName);
	    doc.setContent(content);
	    cont.putDocument(doc);

	    cont.delete();
			
	    // Open the container  
	    cont = mgr.openContainer(containerName2);
			
	    // Get the content
	    System.out.println("Content of the document: "
			       + cont.getDocument(docName).getContentAsString());
			
	    // Clean up
	    compressor.delete();
	    cont.delete();
	    mgr.delete();
			   
	} catch (XmlException e) {
	    System.out.println("Exception: " + e.getMessage());
	}
    }

}

File: dbxml/examples/cxx/misc/compression.cpp
/*
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2004,2009 Oracle. All rights reserved.
 *
 *
 *******
 *
 * Compression API
 *
 * A very simple Berkeley DB XML program that demonstrates
 * compression API usage.
 *
 * This program demonstrates:
 *  Registration of a custom compression class in XmlManager
 *  Compression configuration for XmlContainer
 *  Turn off default compression
 *  Using default compression
 *  Custom compression class implementation
 *
 *  MyCompression is an subclass of XmlCompression. It is a very simple
 *  implementation and not intended to be particularly efficient. As an example,
 *  it uses buffer inverse permutation to simulate the compression process.
 *
 *  Compression only works with whole document containers. User compression must
 *  always be registered with the manager in order to open/create the container.
 *
 */

#include "dbxml/DbXml.hpp"
#include <iostream>

using namespace DbXml;
using namespace std;

/*
 * MyCompression implements the XmlCompression interface: required
 * member functions are compress() and decompress(). The two member functions
 * are required to return tree for success and false for error.
 * This class must be thread-safe in multi-thread environment.
 *
 */
class MyCompression : public XmlCompression
{
	bool compress(XmlTransaction &txn,
			        const XmlData &source,
			        XmlData &dest);
	bool decompress(XmlTransaction &txn,
				   const XmlData &source,
				   XmlData &dest);
};

bool MyCompression::compress(XmlTransaction &txn,
			     const XmlData &source,
			     XmlData &dest)
{
	try {
	// Get the data to compress
	char *pSrc = (char *)source.get_data();
	size_t size = source.get_size();

	// Use inverse permutation to simulate the compression process
	dest.reserve(size);
	char *buf = (char *)dest.get_data();
	for(size_t i=0; i<size; i++)
		buf[i] = pSrc[size-1-i];
	dest.set_size(size);

	} catch (XmlException &xe) {
		cout << "XmlException: " << xe.what() << endl;
		return false;
	}
	return true;
}

bool MyCompression::decompress(XmlTransaction &txn,
			       const XmlData &source,
			       XmlData &dest)
{
	try {
	// Get the data to decompress
	char *pSrc = (char *)source.get_data();
	size_t size = source.get_size();

	// Use inverse permutation to simulate the decompression process
	dest.reserve(size);
	char *buf = (char *)dest.get_data();
	for(size_t i=0; i<size; i++)
		buf[i] = pSrc[size-1-i];
	dest.set_size(size);

	} catch (XmlException &xe) {
		cout << "XmlException: " << xe.what() << endl;
		return false;
	}
	return true;
}

void putDocWithoutDefaultCompression(XmlManager& mgr,
				     const string& containerName,
				     XmlDocument& xdoc,
				     XmlUpdateContext& uc)
{
	// Set the container type as WholedocContainer and turn off
	// the default compression
	XmlContainerConfig contConf;
	contConf.setAllowCreate(true);
	contConf.setContainerType(XmlContainer::WholedocContainer);
	contConf.setCompressionName(XmlContainerConfig::NO_COMPRESSION);

	// Create container
	XmlContainer cont = mgr.createContainer(containerName, contConf);
	
	// Put Document
	cont.putDocument(xdoc, uc);
}

void putDocWithDefaultCompression(XmlManager& mgr,
				  const string& containerName,
				  XmlDocument& xdoc,
				  XmlUpdateContext& uc)
{
	// Set the container type as WholedocContainer and
	// use the default compression
	XmlContainerConfig contConf;
	contConf.setAllowCreate(true);
	contConf.setContainerType(XmlContainer::WholedocContainer);

	// The following line is unnecessary because default compression
	// would take effect if user do not turn off it explicitly.
	contConf.setCompressionName(XmlContainerConfig::DEFAULT_COMPRESSION);

	// Create container
	XmlContainer cont = mgr.createContainer(containerName, contConf);

	// Put Document
	cont.putDocument(xdoc, uc);
}

void putDocWithMyCompression(XmlManager& mgr,
			     const string& containerName,
			     XmlDocument& xdoc,
			     XmlUpdateContext& uc,
			     XmlCompression& myCompression)
{
	// Define an unique name to use for registering the compression
	string compressionName = "myCompression";
	
	// Register custom class
	mgr.registerCompression(compressionName.c_str(), myCompression);

	// Set the container type as WholedocContainer
	// and use the custom compression
	XmlContainerConfig contConf;
	contConf.setAllowCreate(true);
	contConf.setContainerType(XmlContainer::WholedocContainer);
	contConf.setCompressionName(compressionName.c_str());

	// Create container
	XmlContainer cont = mgr.createContainer(containerName, contConf);

	// Put Document
	cont.putDocument(xdoc, uc);
}

string getContent(XmlManager& mgr,
		  const string& containerName,
		  const string& docName)
{
	XmlContainer cont = mgr.openContainer(containerName);
	string content;
	cont.getDocument(docName).getContent(content);
	return content;
}

int main(int argc, char **argv)
{
	string containerName1 = "compressionContainer1.dbxml";
	string containerName2 = "compressionContainer2.dbxml";
	string containerName3 = "compressionContainer3.dbxml";
	string docName = "doc1.xml";
	string content = "<root><a></a></root>";
	try{

	// Create manager and update context. It's good to reuse update context.
	XmlManager mgr;
	XmlUpdateContext uc = mgr.createUpdateContext();

	// Remove the existing container if any
	if( mgr.existsContainer(containerName1) )
		mgr.removeContainer(containerName1);
	if( mgr.existsContainer(containerName2) )
		mgr.removeContainer(containerName2);
	if( mgr.existsContainer(containerName3) )
		mgr.removeContainer(containerName3);

	// Create document and put it into the container without compression.
	XmlDocument xdoc1 = mgr.createDocument();
	xdoc1.setName(docName);
	xdoc1.setContent(content);
	putDocWithoutDefaultCompression(mgr, containerName1, xdoc1, uc);

	// Verify Content
	cout << "Content of the document: "
	     << getContent(mgr, containerName1, xdoc1.getName())
	     << endl;

	// Create document and put it into the container with default compression.
	XmlDocument xdoc2 = mgr.createDocument();
	xdoc2.setName(docName);
	xdoc2.setContent(content);
	putDocWithDefaultCompression(mgr, containerName2, xdoc2, uc);

	// Verify Content
	cout << "Content of the document: "
	     << getContent(mgr, containerName2, xdoc2.getName())
	     << endl;

	// Create an instance of custom compression
	MyCompression myCompression;

	// Create document and put it into the container with custom compression.
	XmlDocument xdoc3 = mgr.createDocument();
	xdoc3.setName(docName);
	xdoc3.setContent(content);
	putDocWithMyCompression(mgr, containerName3, xdoc3, uc, myCompression);

	// Verify Content
	cout << "Content of the document: "
	     << getContent(mgr, containerName3, xdoc3.getName())
	     << endl;

	} catch (XmlException &xe) {
		cout << "XmlException: " << xe.what() << endl;
		return 1;
	}
	return 0;
}
